crates/bevy_animation/src/lib.rs

```diff
@@ -397,196 +493,226 @@ struct AnimationTargetContext<'a> {
+            let Some(animation_graph) = animation_graphs.get(graph_handle) else {
+                return;
+            };
 
-        // Advance the main animation.
-        if let Some(animation_clip) = animation_clips.get(&player.animation.animation_clip) {
-            player
-                .animation
-                .update(time.delta_seconds(), animation_clip.duration);
-        };
+            let AnimationPlayer {
+                ref mut playing_animations,
+                ref blend_weights,
+                ref mut schedule,
+                ..
+            } = *player;
 
-        // Advance transition animations.
-        player.transitions.retain_mut(|transition| {
-            // Decrease weight. Expire the transition if necessary.
-            transition.current_weight -= transition.weight_decline_per_sec * time.delta_seconds();
-            if transition.current_weight <= 0.0 {
-                return false;
-            }
+            // Tick animations, and schedule them.
 
-            if let Some(animation_clip) = animation_clips.get(&transition.animation.animation_clip)
-            {
-                transition
-                    .animation
-                    .update(time.delta_seconds(), animation_clip.duration);
-            };
+            schedule.clear();
 
-            true
+            let mut dfs = animation_graph.dfs();
+            let mut weights = HashMap::with_capacity(animation_graph.graph.node_count());
+
+            while let Some(node_index) = dfs.next(&animation_graph.graph) {
+                let node = &animation_graph[node_index];
+
+                let mut weight = node.weight;
+                for parent in animation_graph
+                    .graph
+                    .neighbors_directed(node_index, Direction::Incoming)
+                {
+                    if let Some(parent_weight) = weights.get(&parent) {
+                        weight *= parent_weight;
+                    }
+                }
+
+                if let Some(playing_animation) = playing_animations.get_mut(&node_index) {
+                    // Tick the animation if necessary.
+                    if !playing_animation.paused {
+                        if let Some(ref clip_handle) = node.clip {
+                            if let Some(clip) = animation_clips.get(clip_handle) {
+                                playing_animation.update(delta_seconds, clip.duration);
+                            }
+                        }
+                    }
+
+                    weight *= playing_animation.weight;
+                } else if let Some(&blend_weight) = blend_weights.get(&node_index) {
+                    weight *= blend_weight;
+                }
+
+                weights.insert(node_index, weight);
+
+                if let (Some(clip), Some(playing_animation)) =
+                    (&node.clip, playing_animations.get(&node_index))
+                {
+                    schedule.push(ScheduledAnimationClip {
```

Comment:
* Can we deduplicate identical clips being added to the ScheduledAnimationClips by accumulating their weights?
* Does this make sense to add clips with a weight of zero to the schedule? 
* Should we terminate the DFS down a branch if we know the weight is already 0?

Reply:
We can't do the first without changing semantics because of the sorting, but if we get rid of the sorting we could.

Reply:
I implemented all these optimizations, and I removed `ScheduledAnimationClips`.